### PR TITLE
PT preload_from_files ext: random init, external part

### PR DESCRIPTION
**Pull request by albertz**
Two features here (maybe we can also leave the individual commits as is):

### 1. `init_for_train=True` and `filename=None` and `pattern`
 
Explicitly specify some parameters which should be randomly initialized (thus `filename=None`) in training.

Note: Currently, if you start training (initial epoch 1), and there is `preload_from_files`, then if that does not cover all model parameters, it will exit with the error: "Missing key(s) in state_dict: ... Any missing key is an error.". I.e. the expectation is, once `preload_from_files` is used, that covers all model parameters.

In further training epochs, when an existing checkpoint is loaded, it will also do the check, but only for those params which are not already in the previous checkpoint. The same logic also applies also for recognition.

So, when you want to use `preload_from_files` for training to preload some of the parameters, and you don't cover all parameters, but you want to leave some parameters randomly initialized, then you need this feature.

### 2. `init_for_train="always"`

Specify some parameters which are always loaded, no matter if it is the initial first training epoch, some later training epoch, or recognition. This also will exclude those parameters from training and exclude those parameters when saving the checkpoint, as it is assumed that they will later again be loaded when they are needed. Or maybe you don't need them anyway for recognition, as they are only used for some auxiliary loss, or for knowledge distillation or so.

### CODE REVIEW

**Comment by Atticus1806** [comment]
>So, when you want to use preload_from_files for training to preload some of the parameters, and you don't cover all parameters, but you want to leave some parameters randomly initialized, then you need this feature.

I am wondering if this is true. I am currently doing training where I preload a model, but it does not contain all the parameters of the network (since I am only loading the teacher). It does not raise an exception, but instead prints all the missing keys. 

The code I am using (removed filename): 
```
"teacher": {
    "filename": ...,
    "ignore_missing": False,
    "ignore_params_prefixes": [],
    "init_for_train": True,
    "prefix": "teacher.",
}
```

Or what is the relation to `filename=None`? Where from would I preload without a file? 
I am probably missing the point of this PR?

**Comment by albertz** [comment]
I mean this exception, which I get without this:
```python
if missing_keys:
    raise Exception(
        "\n".join(
            [
                f"While loading model {filename}, preload_from_files {bool(preload_from_files)}:",
                "Unexpected key(s) in state_dict: " + ", ".join(map(repr, sorted(unexpected_keys))),
                "Missing key(s) in state_dict: " + ", ".join(map(repr, sorted(missing_keys))),
                "Any missing key is an error.",
            ]
        )
    )
```
You don't get this because you have a `prefix` here, and the preload is *complete* for that prefix.

I have some preload which is *not complete*. Specifically, I have:
```python
preload_from_files = {
    ...,  # also some other things
    "03_encoder": {
        "init_for_train": "always",  # or True
        # this checkpoint here does not cover the full model, just the encoder
        "filename": enc.checkpoint,
        # no prefix here
    },
}
```

> Or what is the relation to `filename=None`? Where from would I preload without a file?

With `filename=None`, it would load from nowhere. It would not preload from anything but instead randomly initialize the parameters. But that will then not have `missing_keys`.

File: returnn/torch/util/module.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    parameter = getattr(module, parameter_name)
+    if not isinstance(parameter, torch.nn.Parameter):
+        raise ValueError(f"{parameter_name} is not a torch.nn.Parameter, got type {type(parameter).__name__}")
+    delattr(module, parameter_name)
+    parameter.requires_grad = False
+    module.register_buffer(parameter_name, parameter, persistent=persistent)
+
+
+def unregister_module_or_parameter_or_buffer(module: torch.nn.Module, attrib_name: str, *, deep: bool = True):
```
Comment (NeoLegends) [inline returnn/torch/util/module.py:46]: Is this function relevant for the PR?

**Review by NeoLegends — APPROVED** [review]

File: returnn/torch/util/module.py (hunk 1 above)
Comment (albertz) [reply returnn/torch/util/module.py]: Ah yea I think I used it in an earlier version. Can be removed.

File: returnn/torch/engine.py [hunk 2]
```diff
@@ -913,9 +951,16 @@ def _save_model(self):
         tmp_filename = filename + ".tmp_write"
         if os.path.exists(tmp_filename):
             os.unlink(tmp_filename)
+        state_dict = self._pt_model.state_dict()
+        if self._ignore_param_set:
+            # Do some extra check that we don't save the ignored parameters.
```
Comment (NeoLegends) [inline returnn/torch/engine.py:956]: I was wondering if this would cause downstream ergonomics issues if the resulting model was used outside of RETURNN, where the preloading logic was less handily available, e.g. in ONNX, and if you then do suddenly end up depending on their values. I suppose this is a weird edge case though, because if so, just load these parameters normally in training and freeze them.

Comment (albertz) [reply returnn/torch/engine.py:956]: Yea I would say this is up to the user. RETURNN should not restrict the user to do what he/she wants.
